stdlib/public/CommandLineSupport/CommandLine.cpp

```diff
@@ -46,120 +47,206 @@ void _swift_stdlib_overrideUnsafeArgvArgc(char **argv, int argc) {
+  /// 
+  /// The result of this function is @em not owned by the caller and should
+  /// persist for the lifetime of the process.
+  static char **getUnsafeArgvArgc(int *outArgLen);
+
+  /// A platform-specific function that enumerates the contents of @c argv
+  /// one argument at a time.
+  /// 
+  /// @a body is a function that takes two arguments:
+  /// 
+  /// - The first argument is the value of @c argc if it can be readily
+  ///   computed, or @c -1 otherwise.
+  /// - The second argument is the element of @c argv being enumerated. The
+  ///   caller makes a copy of this string. The implementation should not
+  ///   enumerate the trailing @c nullptr required by the C standard.
+  /// 
+  /// Callers should call @c getUnsafeArgvArgc() before calling this function
+  /// in case a fast path is available. If that function is implemented on this
+  /// platform, then this function does not need to be implemented.
+  template <typename F>
+  static void enumerateUnsafeArgv(const F& body);
+}
 
 SWIFT_CC(swift) SWIFT_RUNTIME_STDLIB_INTERNAL
 char **_swift_stdlib_getUnsafeArgvArgc(int *outArgLen) {
   assert(outArgLen != nullptr);
 
-  if (_swift_stdlib_ProcessOverrideUnsafeArgv) {
+  // Check the override before doing any platform-specific work.
+  if (SWIFT_UNLIKELY(_swift_stdlib_ProcessOverrideUnsafeArgv)) {
     *outArgLen = _swift_stdlib_ProcessOverrideUnsafeArgc;
     return _swift_stdlib_ProcessOverrideUnsafeArgv;
   }
 
+  // Try the platform-specific fast path that avoids heap (re)allocation. Not
+  // all platforms implement this function.
+  if (auto argv = swift::getUnsafeArgvArgc(outArgLen)) {
+    return argv;
+  }
+
+  // Take a slower path wherein we construct the argv array one element at a
+  // time. If the enumeration function can provide us with a hint for argc, we
+  // can avoid calling realloc() more than once.
+  // 
+  // NOTE: It is intentional that we do not simply use std::vector here. STL
+  // collections may call operator new() which can be overridden by client code
+  // and that client code could call back into Swift.
+  int maxArgc = 0;
+  int argc = 0;
+  char **argv = nullptr;
+  swift::enumerateUnsafeArgv([&] (int maxArgcHint, const char *arg) {
+    if (argc >= maxArgc) {
+      if (maxArgcHint > maxArgc) {
+        // The platform was able to cheaply get argc, so use the
+        // provided value.
+        maxArgc = maxArgcHint;
+      } else {
+        // The platform could not tell us argc, so grow the argument array by
+        // small amounts. We assume here that the command line doesn't usually
+        // have many arguments, so a small/linear growth factor should be safe.
+        maxArgc += 16;
+      }
+
+      // Reallocate (or initially allocate) the argv buffer. Overallocate by
+      // one element to allow for a trailing nullptr.
+      size_t argvSize = sizeof(char *) * (maxArgc + 1);
+      argv = reinterpret_cast<char **>(realloc(argv, argvSize));
```

Comment:
I suspect that I am missing something, but why?  Why do all this work on the slow path?  Why not just use a `std::vector<char *>` statically and grow the backing store, and just return that subsequently and populate it the first time?

Reply:
`std::vector` uses `operator new` under the covers. That function can be overridden, as you know. So what happens when you override `operator new` and call out to Swift? 💥

Reply:
The corollary to that is that we could write a `std::allocator` based on `swift_slowAlloc()` (or whatevs) but… this is somehow _less_ complicated than rolling your own allocator.